scripts/data/src/org/akvo/gae/remoteapi/DataStoreWithUnilog.java

```diff
@@ -0,0 +1,265 @@
+    private boolean newEventsPushedToEventQueue;
+
+    DataStoreWithUnilog(DatastoreService delegate, String orgId) {
+        this.delegate = delegate;
+        this.orgId = orgId;
+    }
+
+    @Override
+    public Key put(Entity entity) {
+        Key put = delegate.put(entity);
+        storeEntityUpdate(Stream.of(entity));
+        return put;
+    }
+
+    @Override
+    public Key put(Transaction transaction, Entity entity) {
+        Key put = delegate.put(transaction, entity);
+        storeEntityUpdate(Stream.of(entity));
+        return put;
+    }
+
+    @Override
+    public List<Key> put(Iterable<Entity> iterable) {
+        List<Key> put = delegate.put(iterable);
+        Stream<Entity> entities = asStream(iterable);
+        storeEntityUpdate(entities);
+        return put;
+    }
+
+    @Override
+    public List<Key> put(Transaction transaction, Iterable<Entity> iterable) {
+        List<Key> put = delegate.put(transaction, iterable);
+        storeEntityUpdate(asStream(iterable));
+        return put;
+    }
+
+    @Override
+    public void delete(Key... keys) {
+        delegate.delete(keys);
+        storeDeleteEvents(Arrays.stream(keys));
+    }
+
+    @Override
+    public void delete(Transaction transaction, Key... keys) {
+        delegate.delete(transaction, keys);
+        storeDeleteEvents(Arrays.stream(keys));
+    }
+
+    @Override
+    public void delete(Iterable<Key> iterable) {
+        delegate.delete(iterable);
+        storeDeleteEvents(asStream(iterable));
+    }
+
+    @Override
+    public void delete(Transaction transaction, Iterable<Key> iterable) {
+        delegate.delete(transaction, iterable);
+        storeDeleteEvents(asStream(iterable));
+    }
+
+    private <T> Stream<T> asStream(Iterable<T> iterable) {
+        return StreamSupport.stream(iterable.spliterator(), false);
+    }
+
+    private void storeEntityUpdate(Stream<Entity> entities) {
+        storeInEventQueue(entities.map(this::event));
+    }
+
+    private void storeInEventQueue(Stream<Entity> entityStream) {
+        List<Entity> events = entityStream
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
+        if (!events.isEmpty()) {
+            newEventsPushedToEventQueue = true;
+            delegate.put(events);
+        }
+    }
+
+    private Map<String, Object> manualFixEventSource() {
+        return EventUtils.newSource("manualDataFix");
+    }
+
+    private Entity event(Entity entity) {
+        String kind = entity.getKey().getKind();
+        final EventTypes eventTypes = EventUtils
+                .getEventAndActionType(kind);
+
+        if (eventTypes != null) {
+            Map<String, Object> event = updateOrCreateEvent(entity, eventTypes);
+            try {
+                return EventUtils.createEventLogEntity(event, new Date());
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        }
+        return entity;
```

Comment:
Does this mean if no event type is found then the entity would get stored as is in the event queue?

Reply:
Correct. It is a bug. Fixed. 

Good catch!